### PR TITLE
[RW-7801][risk=low] Added Email for CT Reminder and Updated RT Reminders

### PR DESCRIPTION
**Pull request by evrii**
Description:
RT Expired:
image

RT Expiring:
image

---
**PR checklist**

- [ ] This PR meets the Acceptance Criteria in the JIRA story
- [ ] The JIRA story has been moved to Dev Review
- [ ] This PR includes appropriate unit tests
- [ ] I have added explanatory comments where the logic is not obvious
- [ ] I have run and tested this change locally, and my testing process is described here
- [ ] If this includes a new feature flag, I have created and linked new JIRA tickets to (a) turn on the feature flag and (b) remove it later
- [ ] If this includes an API change, I have run the E2E tests on this change against my local server with [yarn test-local](https://github.com/all-of-us/workbench/blob/master/e2e/README.md#examples) because this PR won't be covered by the CircleCI tests 
- [ ] If this includes a UI change, I have taken screen recordings or screenshots of the new behavior and notified the PO and UX designer in [Slack](https://pmi-engteam.slack.com/archives/C02MWP2RN5P)
- [ ] If this change impacts deployment safety (e.g. removing/altering APIs which are in use) I have documented these in the description
- [ ] If this includes an API change, I have updated the appropriate Swagger definitions and updated the appropriate API consumers
  * AoU UI
  * [Perf tests](https://github.com/broadinstitute/mcnulty/blob/develop/src/test/scala/services/AoU.scala)
  * [Researcher Directory export](https://github.com/all-of-us/workbench/wiki/Researcher-Directory-(RDR-export))
  * Cron tasks - for Offline*Controllers
  * SumoLogic - for EgressAlert

### CODE REVIEW

File: api/src/main/java/org/pmiops/workbench/api/OfflineUserController.java [hunk 1]
```diff
@@ -37,7 +41,13 @@ public ResponseEntity<Void> synchronizeUserAccess() {
 
   @Override
   public ResponseEntity<Void> sendAccessExpirationEmails() {
-    userService.getAllUsers().forEach(userService::maybeSendAccessExpirationEmail);
+    userService.getAllUsers().forEach(this::maybeSendAccessExpirationEmails);
+
     return ResponseEntity.noContent().build();
   }
+
+  private void maybeSendAccessExpirationEmails(DbUser user) {
```
Comment (evrii) [inline api/src/main/java/org/pmiops/workbench/api/OfflineUserController.java:49]: Should this method be in the service instead?

File: api/src/main/resources/emails/access_expired/content.mjml [hunk 2]
```diff
@@ -0,0 +1,68 @@
+<mjml>
```
Comment (jmthibault79) [inline api/src/main/resources/emails/access_expired/content.mjml:1]: could you add a short readme or wiki entry about how to turn these into HTML ?

File: api/src/main/java/org/pmiops/workbench/api/OfflineUserController.java (hunk 1 above)
Comment (evrii) [reply api/src/main/java/org/pmiops/workbench/api/OfflineUserController.java]: Moved it. Thanks for the second set of eyes.

File: api/src/main/java/org/pmiops/workbench/mail/MailServiceImpl.java [hunk 3]
```diff
@@ -221,44 +226,74 @@ public void alertUserRegisteredTierWarningThreshold(
             formatCentralTime(expirationTime));
     log.info(logMsg);
 
+    String resource;
+
+    switch (tierShortName) {
```
[the remark this answers is not in the document]
Comment (evrii) [reply api/src/main/java/org/pmiops/workbench/mail/MailServiceImpl.java:231]: Yeah, I hadn't had a chance to change the file name before I left. From what I can remember, I think this was a placeholder until I could get everything updated.

[the remark this answers is not in the document]
Comment (evrii) [reply api/src/main/java/org/pmiops/workbench/mail/MailServiceImpl.java]: Updated this to use the updated constant.

File: api/src/main/java/org/pmiops/workbench/db/dao/UserService.java [hunk 4]
```diff
@@ -152,7 +152,7 @@ DbUser syncComplianceTrainingStatusV2(DbUser user, Agent agent)
   DbUser confirmPublications();
 
   /** Send an Access Renewal Expiration or Warning email to the user, if appropriate */
-  void maybeSendAccessExpirationEmail(DbUser user);
+  void maybeSendAccessTierExpirationEmails(DbUser user);
```
Comment (dmohs) [inline api/src/main/java/org/pmiops/workbench/db/dao/UserService.java:155]: Good rename.

File: api/src/main/java/org/pmiops/workbench/db/dao/UserServiceImpl.java [hunk 5]
```diff
@@ -804,13 +810,11 @@ public DbUser confirmPublications() {
     return accessSyncService.updateUserAccessTiers(dbUser, Agent.asUser(dbUser));
   }
 
-  /** Send an Access Renewal Expiration or Warning email to the user, if appropriate */
+  /** Send Access Renewal Expiration or Warning emails to the user for all tiers, if appropriate */
```
Comment (dmohs) [inline api/src/main/java/org/pmiops/workbench/db/dao/UserServiceImpl.java:813]: This comment increases the vertical space consumed by 25%. Consider whether or not it is adding enough value to pay for that space.

File: api/src/main/java/org/pmiops/workbench/db/dao/UserServiceImpl.java [hunk 6]
```diff
@@ -880,4 +903,15 @@ private void maybeSendRegisteredTierExpirationEmail(DbUser user, Timestamp expir
   private Timestamp clockNow() {
     return new Timestamp(clock.instant().toEpochMilli());
   }
+
+  /**
+   * Send an Access Renewal Expiration or Warning email to the user for a specified tier, if
+   * appropriate
+   */
```
Comment (dmohs) [inline api/src/main/java/org/pmiops/workbench/db/dao/UserServiceImpl.java:910]: Same consideration for this comment as above.

**Review by dmohs — APPROVED** [review]

File: api/src/main/java/org/pmiops/workbench/db/dao/UserServiceImpl.java [hunk 7]
```diff
@@ -880,4 +903,12 @@ private void maybeSendRegisteredTierExpirationEmail(DbUser user, Timestamp expir
   private Timestamp clockNow() {
     return new Timestamp(clock.instant().toEpochMilli());
   }
+
+  // Send an Access Renewal Expiration or Warning email to the user, if appropriate
+  private void maybeSendAccessTierExpirationEmail(DbUser user, String tierShortName) {
+    final Optional<Timestamp> accessTierExpiration =
+        getTierExpirationForEmails(user, tierShortName);
+    accessTierExpiration.ifPresent(
```
Comment (jmthibault79) [inline api/src/main/java/org/pmiops/workbench/db/dao/UserServiceImpl.java:911]: optional: could do this instead
```
getTierExpirationForEmails(user, tierShortName).ifPresent(...)
```

File: api/src/test/java/org/pmiops/workbench/access/UserServiceAccessTest.java [hunk 8]
```diff
@@ -494,16 +501,48 @@ public void test_maybeSendAccessExpirationEmail_expiring_1() throws MessagingExc
     accessModuleService.updateCompletionTime(
         dbUser, DbAccessModuleName.RT_COMPLIANCE_TRAINING, willExpireAfter(oneDayPlusSome));
 
-    userService.maybeSendAccessExpirationEmail(dbUser);
+    userService.maybeSendAccessTierExpirationEmails(dbUser);
+
+    verify(mailService)
+        .alertUserAccessTierWarningThreshold(dbUser, 1, expirationTime, REGISTERED_TIER_SHORT_NAME);
+
+    // Control Tier Threshold Email is sent because RT Compliance Training is a requirement for CT
```
Comment (jmthibault79) [inline api/src/test/java/org/pmiops/workbench/access/UserServiceAccessTest.java:509]: oh.

I realize this is pretty late in the game, but... is that what we want?

File: api/src/test/java/org/pmiops/workbench/access/UserServiceAccessTest.java [hunk 9]
```diff
@@ -734,16 +783,49 @@ public void test_maybeSendAccessExpirationEmail_expired() throws MessagingExcept
...
         dbUser, DbAccessModuleName.RT_COMPLIANCE_TRAINING, expiredBy(oneHour));
 
-    userService.maybeSendAccessExpirationEmail(dbUser);
+    userService.maybeSendAccessTierExpirationEmails(dbUser);
+
+    // Registered Tier Expiration Email is sent
+    verify(mailService)
+        .alertUserAccessTierExpiration(dbUser, expirationTime, REGISTERED_TIER_SHORT_NAME);
+
+    // Controlled Tier Expiration Email is sent because RT Compliance is a requirement for CT Tier.
```
Comment (jmthibault79) [inline api/src/test/java/org/pmiops/workbench/access/UserServiceAccessTest.java:792]: It doesn't look like the user previously had CT access in this case, so an email should not be sent.

File: api/src/test/java/org/pmiops/workbench/utils/TestMockFactory.java [hunk 10]
```diff
@@ -73,7 +73,7 @@ public class TestMockFactory {
    * them bypassable and non-expirable. So we should default to this, unless required for product
    * reasons.
    *
-   * <p>RAS_LOGIN_GOV and CT_COMPLIANCE_TRAINING are bypassable and non-expirable for this reason.
```
Comment (jmthibault79) [inline api/src/test/java/org/pmiops/workbench/utils/TestMockFactory.java:76]: did you find that CT training is expirable in the DB?

File: api/src/main/resources/emails/tier_access_threshold/content.mjml [hunk 11]
```diff
@@ -0,0 +1,68 @@
+<mjml>
+  <mj-head>
+    <mj-attributes>
+      <mj-section padding="0 0 10px 0" />
+      <mj-text font-size=24 line-height=1.5 font-family="'Montserrat', sans-serif"/>
```
Comment (jmthibault79) [inline api/src/main/resources/emails/tier_access_threshold/content.mjml:5]: mjml gave me this error 
```
Line 5 of /Users/thibault/src/aou/api/src/main/resources/emails/tier_access_threshold/tier_access_threshold/content.mjml (mj-text) — Attribute font-size has invalid value: 24 for type Unit, only accepts (px) units and 1 value(s)
```

File: api/src/main/resources/emails/README.md [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+[MJML](https://mjml.io/) is a node package that allows you to write emails in
+a [custom markup language](https://documentation.mjml.io/) that can be converted into HTML that will
+be rendered [consistently across many email clients](https://mjml.io/faq#email-clients)
+
+## Installation
+
+We do not currently have the infrastructure to support node packages in our backend, so you will
+need to install MJML locally:
+
+```npm install -g mjml@4.13.0```
```
Comment (jmthibault79) [inline api/src/main/resources/emails/README.md:19]: I had to use sudo for this, not sure if worth mentioning.  Probably not.

File: api/src/main/resources/emails/README.md [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+*Note: We do not currently have a package.json for our api, but for consistency, please use version
+4.13.0 of MJML.*
+
+## Helpful Commands
+
+There are two commands that you should be aware of.
+
+First, in order to convert an mjml file to an html file, you can run:
+
+```mjml index.mjml index.html```
```
Comment (jmthibault79) [inline api/src/main/resources/emails/README.md:30]: This command didn't work for me as-is.  I needed to redirect output:
```
mjml index.mjml > index.html
```

**Review by jmthibault79 — COMMENTED** [review]
A few concerns but it is looking good overall

File: api/src/test/java/org/pmiops/workbench/access/UserServiceAccessTest.java (hunk 8 above)
Comment (jmthibault79) [reply api/src/test/java/org/pmiops/workbench/access/UserServiceAccessTest.java:509]: I do remember discussing the idea of unified logic (so we only send one email) and rejecting that for the first iteration.  So yeah I think this meets that agreement.

I didn't consider the case where CT is expiring for *strictly RT-related reasons* so I am less happy with this plan now.  But let's keep this as-is, and get back to the more polished version in the near future.

Comment (evrii) [reply api/src/test/java/org/pmiops/workbench/access/UserServiceAccessTest.java:509]: Yeah, I don't think I fully considered that case when we first started either. Your suggestion sounds like a good approach to me.

**Review by NehaBroad — APPROVED** [review]

File: api/src/main/java/org/pmiops/workbench/db/dao/UserServiceImpl.java [hunk 14]
```diff
@@ -880,4 +890,28 @@ private void maybeSendRegisteredTierExpirationEmail(DbUser user, Timestamp expir
...
+  // Send an Access Renewal Expiration or Warning email to the user, if appropriate
+  private void maybeSendAccessTierExpirationEmail(DbUser user, String tierShortName) {
+    List<DbAccessModuleName> requiredModules;
+
+    switch (tierShortName) {
+      case REGISTERED_TIER_SHORT_NAME:
+        requiredModules = REQUIRED_MODULES_FOR_REGISTERED_TIER;
+        break;
+      case CONTROLLED_TIER_SHORT_NAME:
+        requiredModules =
```
Comment (jmthibault79) [inline api/src/main/java/org/pmiops/workbench/db/dao/UserServiceImpl.java:903]: lgtm.  We can revisit this logic as we discussed, in a later PR

File: api/src/main/java/org/pmiops/workbench/db/dao/UserServiceImpl.java [hunk 15]
```diff
@@ -831,12 +835,17 @@ public void signOut(DbUser user) {
    * <p>Note that this method may return EMPTY for both valid and invalid users, so this method
    * SHOULD NOT BE USED FOR ACCESS DECISIONS.
    */
-  private Optional<Timestamp> getRegisteredTierExpirationForEmails(DbUser user) {
+  private Optional<Timestamp> getTierExpirationForEmails(
```
[the remark this answers is not in the document]
Comment (evrii) [reply api/src/main/java/org/pmiops/workbench/db/dao/UserServiceImpl.java:838]: What are you thinking for an update? Is it the expiration times bit? I think that test_maybeSendAccessTierExpirationEmails_expired_but_missing shows that if one of the requiredModules does have a null expiration time, the tier is considered incomplete and the email is not sent. In that case, I think this comment is still valid.

File: api/src/test/java/org/pmiops/workbench/utils/TestMockFactory.java (hunk 10 above)
Comment (evrii) [reply api/src/test/java/org/pmiops/workbench/utils/TestMockFactory.java:76]: Yeah, local and prod both have expirable CT Compliance Training

File: api/src/main/java/org/pmiops/workbench/db/dao/UserServiceImpl.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (jmthibault79) [reply api/src/main/java/org/pmiops/workbench/db/dao/UserServiceImpl.java:838]: I mean that it refers to RT and it's now tier-generic

[the remark this answers is not in the document]
Comment (evrii) [reply api/src/main/java/org/pmiops/workbench/db/dao/UserServiceImpl.java:838]: Ahh, I see it now. Thanks!

**Pull request merged**